File: app/src/main/java/com/kidozh/discuzhub/utilities/timeDisplayUtils.java
package com.kidozh.discuzhub.utilities;

import android.content.Context;
import android.util.Log;


import com.kidozh.discuzhub.R;

import java.text.DateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.Locale;

public class timeDisplayUtils {
    private static String TAG = timeDisplayUtils.class.getSimpleName();
    private Context mContext;
    public timeDisplayUtils(Context mContext){
        this.mContext = mContext;
    }

    public static Calendar dateToCalendar(Date date) {
        Calendar calendar = Calendar.getInstance();
        calendar.setTime(date);
        return calendar;
    }

    public static String getLocalePastTimeString(Context mContext,Date date){
        Date now = new Date();
        Calendar nowCalendar = Calendar.getInstance();
        Calendar dateCalendar = dateToCalendar(date);
        int nowDay = nowCalendar.get(Calendar.DAY_OF_YEAR);
        int dateDay = dateCalendar.get(Calendar.DAY_OF_YEAR);
        int nowYear = nowCalendar.get(Calendar.YEAR);
        int dateYear = dateCalendar.get(Calendar.YEAR);
        DateFormat timeFormat = DateFormat.getTimeInstance(DateFormat.SHORT, Locale.getDefault());
        String timeString = timeFormat.format(date);
        long timeMillsecondsInterval = now.getTime() - date.getTime();
        // handle ago
        if(timeMillsecondsInterval >= 0){
            if(timeMillsecondsInterval < 1000* 60 ){
                return mContext.getString(R.string.date_just_now);
            }
            else if (timeMillsecondsInterval < 1000* 60 * 60 ){
                return String.format(mContext.getString(R.string.date_minutes_before), String.valueOf((int) timeMillsecondsInterval/(60*1000)));
            }
            else if (timeMillsecondsInterval < 1000* 60 * 60 * 24 ){
                return String.format(mContext.getString(R.string.date_hours_before), String.valueOf((int) timeMillsecondsInterval/(60*60*1000)));
            }
            else if(nowDay == dateDay){

                String timeTemplate = mContext.getString(R.string.time_template_today_time);

                return String.format(timeTemplate,timeString);
            }
            else if(nowDay - dateDay > 0 && nowYear == dateYear) {
                int intervalDay = nowDay - dateDay;
                if(intervalDay == 1){
                    String timeTemplate = mContext.getString(R.string.time_template_yesterday_time);
                    return String.format(timeTemplate,timeString);
                }
                else if(intervalDay < 10) {
                    String timeTemplate = mContext.getString(R.string.time_template_days_ago);
                    return String.format(timeTemplate,intervalDay,timeString);
                }
                else {
                    DateFormat df = DateFormat.getDateTimeInstance(DateFormat.SHORT, DateFormat.MEDIUM, Locale.getDefault());
                    return df.format(date);
                }

            }
            else {
                DateFormat df = DateFormat.getDateTimeInstance(DateFormat.SHORT, DateFormat.MEDIUM, Locale.getDefault());
                return df.format(date);
            }
        }
        else {
            // future
            // reverse it
            timeMillsecondsInterval = -timeMillsecondsInterval;
            Log.d(TAG,"Future time " + timeMillsecondsInterval);
            if(timeMillsecondsInterval < 1000* 60 ){
                return mContext.getString(R.string.date_right_away);
            }
            else if (timeMillsecondsInterval < 1000* 60 * 60 ){
                return mContext.getString(R.string.date_in_minutes,(int) timeMillsecondsInterval/(60*1000));
            }
            else if (timeMillsecondsInterval < 1000* 60 * 60 * 24 ){
                return mContext.getString(R.string.date_in_hours,(int) timeMillsecondsInterval/(60*60*1000));
            }
            else if(nowDay == dateDay){

                String timeTemplate = mContext.getString(R.string.time_template_today_time);

                return String.format(timeTemplate,timeString);
            }
            else if(dateDay - nowDay > 0 && nowYear == dateYear) {
                int intervalDay = dateDay - nowDay;
                if(intervalDay == 1){
                    return mContext.getString(R.string.date_tomorrow_time_template,timeString);
                }
                else if(intervalDay < 10) {
                    return mContext.getString(R.string.date_in_days,intervalDay,timeString);
                }
                else {
                    DateFormat df = DateFormat.getDateTimeInstance(DateFormat.SHORT, DateFormat.MEDIUM, Locale.getDefault());
                    return df.format(date);
                }

            }
            else {
                DateFormat df = DateFormat.getDateTimeInstance(DateFormat.SHORT, DateFormat.MEDIUM, Locale.getDefault());
                return df.format(date);
            }
        }

    }
}

File: app/src/main/java/com/kidozh/discuzhub/utilities/TimeDisplayUtils.kt
package com.kidozh.discuzhub.utilities

import android.content.Context
import android.util.Log
import com.kidozh.discuzhub.R
import java.text.DateFormat
import java.util.*

class TimeDisplayUtils(private val mContext: Context) {
    companion object {
        private val TAG = TimeDisplayUtils::class.java.simpleName
        fun dateToCalendar(date: Date?): Calendar {
            val calendar = Calendar.getInstance()
            calendar.time = date
            return calendar
        }

        @JvmStatic
        fun getLocalePastTimeString(mContext: Context, date: Date): String {
            val now = Date()
            val nowCalendar = Calendar.getInstance()
            val dateCalendar = dateToCalendar(date)
            val nowDay = nowCalendar[Calendar.DAY_OF_YEAR]
            val dateDay = dateCalendar[Calendar.DAY_OF_YEAR]
            val nowYear = nowCalendar[Calendar.YEAR]
            val dateYear = dateCalendar[Calendar.YEAR]
            val timeFormat = DateFormat.getTimeInstance(DateFormat.SHORT, Locale.getDefault())
            val timeString = timeFormat.format(date)
            var timeMillsecondsInterval = now.time - date.time
            // handle ago
            return if (timeMillsecondsInterval >= 0) {
                if (timeMillsecondsInterval < 1000 * 60) {
                    mContext.getString(R.string.date_just_now)
                } else if (timeMillsecondsInterval < 1000 * 60 * 60) {
                    String.format(mContext.getString(R.string.date_minutes_before), (timeMillsecondsInterval.toInt() / (60 * 1000)).toString())
                } else if (timeMillsecondsInterval < 1000 * 60 * 60 * 24) {
                    String.format(mContext.getString(R.string.date_hours_before), (timeMillsecondsInterval.toInt() / (60 * 60 * 1000)).toString())
                } else if (nowDay == dateDay) {
                    val timeTemplate = mContext.getString(R.string.time_template_today_time)
                    String.format(timeTemplate, timeString)
                } else if (nowDay - dateDay > 0 && nowYear == dateYear) {
                    val intervalDay = nowDay - dateDay
                    if (intervalDay == 1) {
                        val timeTemplate = mContext.getString(R.string.time_template_yesterday_time)
                        String.format(timeTemplate, timeString)
                    } else if (intervalDay < 10) {
                        val timeTemplate = mContext.getString(R.string.time_template_days_ago)
                        String.format(timeTemplate, intervalDay, timeString)
                    } else {
                        val df = DateFormat.getDateTimeInstance(DateFormat.SHORT, DateFormat.MEDIUM, Locale.getDefault())
                        df.format(date)
                    }
                } else {
                    val df = DateFormat.getDateTimeInstance(DateFormat.SHORT, DateFormat.MEDIUM, Locale.getDefault())
                    df.format(date)
                }
            } else {
                // future
                // reverse it
                timeMillsecondsInterval = -timeMillsecondsInterval
                Log.d(TAG, "Future time $timeMillsecondsInterval")
                if (timeMillsecondsInterval < 1000 * 60) {
                    mContext.getString(R.string.date_right_away)
                } else if (timeMillsecondsInterval < 1000 * 60 * 60) {
                    mContext.getString(R.string.date_in_minutes, timeMillsecondsInterval.toInt() / (60 * 1000))
                } else if (timeMillsecondsInterval < 1000 * 60 * 60 * 24) {
                    mContext.getString(R.string.date_in_hours, timeMillsecondsInterval.toInt() / (60 * 60 * 1000))
                } else if (nowDay == dateDay) {
                    val timeTemplate = mContext.getString(R.string.time_template_today_time)
                    String.format(timeTemplate, timeString)
                } else if (dateDay - nowDay > 0 && nowYear == dateYear) {
                    val intervalDay = dateDay - nowDay
                    if (intervalDay == 1) {
                        mContext.getString(R.string.date_tomorrow_time_template, timeString)
                    } else if (intervalDay < 10) {
                        mContext.getString(R.string.date_in_days, intervalDay, timeString)
                    } else {
                        val df = DateFormat.getDateTimeInstance(DateFormat.SHORT, DateFormat.MEDIUM, Locale.getDefault())
                        df.format(date)
                    }
                } else {
                    val df = DateFormat.getDateTimeInstance(DateFormat.SHORT, DateFormat.MEDIUM, Locale.getDefault())
                    df.format(date)
                }
            }
        }
    }
}